cdap-docs/reference-manual/source/release-notes.rst

```diff
@@ -17,12 +17,634 @@
+  directly throwing a ``TransactionFailureException`` and at the same time handle the
+  previous style of the ``TransactionFailureException`` being wrapped in a
+  ``RuntimeException``.
+
+  Code that is only used in CDAP 4.0.0 and higher can use a simpler version of this::
+
+      @Override
+      public void run() {
+        try {
+          getContext().execute(new TxRunnable() {
+            @Override
+            public void run(DatasetContext context) throws Exception {
+              if (getContext().getRuntimeArguments().containsKey("fail")) {
+                throw new RuntimeException("fail");
+              }
+            }
+          });
+        } catch (TransactionFailureException e) {
+          ...
+        }
+      }
+    }
+  
+- :cask-issue:`CDAP-7544` - The :ref:`Metadata HTTP RESTful API
+  <http-restful-api-metadata-searching>` has been modified to support sorting and
+  pagination. To do so, the API now uses additional parameters |---| ``sort``, ``offset``,
+  ``limit``, ``numCursors``, and ``cursor`` |---| and the format of the results
+  returned when searching has changed. Whereas previous to CDAP 4.0.0 the API returned
+  results as a list of results, the API now returns the results as a field in a JSON object.
+
+- :cask-issue:`CDAP-7796` - Two properties are changing in version 4.0.0 of the CSD:
+
+  - ``log.saver.run.memory.megs`` is replaced with ``log.saver.container.memory.mb``
+  - ``log.saver.run.num.cores`` is replaced with ``log.saver.container.num.cores``
+
+  Anyone who has modified these properties in previous versions will have to update them
+  after upgrading.
+
+Deprecated and Removed Features
+-------------------------------
+
+- :cask-issue:`CDAP-5246` - Removed the deprecated Kafka feed for metadata updates. Users
+  should instead subscribe to the CDAP Audit feed, which contains metadata update
+  notifications in messages with audit type ``METADATA_CHANGE``.
+
+- :cask-issue:`CDAP-6862` - Deprecated "log.saver.run.memory.megs" and
+  "log.saver.run.num.cores", in favor of "log.saver.container.memory.mb" and
+  "log.saver.container.num.cores", respectively.
+
+- :cask-issue:`CDAP-7475` - Removes deprecated methods ``setInputDataset()``,
+  ``setOutputDataset()``, and ``useStreamInput()`` from the MapReduce API, and related
+  methods from the MapReduceContext.
+
+- :cask-issue:`CDAP-7718` - Removed the deprecated ``StreamBatchReadable`` class.
+
+- :cask-issue:`CDAP-7127` - The deprecated CDAP Explore service instance property has been
+  removed.
+
+- :cask-issue:`CDAP-7205` - Removes the deprecated ``useDatasets()`` method from API and
+  documentation.
+
+- :cask-issue:`CDAP-7563` - Removed the usage of deprecated methods from examples.
+
+- :cask-issue:`HYDRATOR-1094` - Removed the deprecated
+  ``cdap-etl-batch-source-archetype``, ``cdap-etl-batch-sink-archetype``, and
+  ``cdap-etl-transform-archetype`` in favor of the ``cdap-data-pipeline-plugins-archetype``.
+
+
+New Features
```

Comment:
Main features should come first. 
1. Cask Market
2. Wrangler
3. Transactional Messaging system
4. Dynamic Log Level
5. Operational Stats
6. New versions of Distros supported
7. New plugins added